controller/nodegroup.go

```diff
@@ -0,0 +1,234 @@
+	}, nil
+}
+
+func createNewLaunchTemplateVersion(launchTemplateID string, group eksv1.NodeGroup, ec2Service *ec2.EC2) (*eksv1.LaunchTemplate, error) {
+	launchTemplateVersionInput := &ec2.CreateLaunchTemplateVersionInput{
+		LaunchTemplateData: buildLaunchTemplateData(group),
+		LaunchTemplateId:   aws.String(launchTemplateID),
+	}
+
+	awsLaunchTemplateOutput, err := ec2Service.CreateLaunchTemplateVersion(launchTemplateVersionInput)
+	if err != nil {
+		return nil, err
+	}
+
+	return &eksv1.LaunchTemplate{
+		Name:    awsLaunchTemplateOutput.LaunchTemplateVersion.LaunchTemplateName,
+		ID:      awsLaunchTemplateOutput.LaunchTemplateVersion.LaunchTemplateId,
+		Version: awsLaunchTemplateOutput.LaunchTemplateVersion.VersionNumber,
+	}, nil
+}
+
+func buildLaunchTemplateData(group eksv1.NodeGroup) *ec2.RequestLaunchTemplateData {
+	launchTemplateData := &ec2.RequestLaunchTemplateData{
+		ImageId:  group.ImageID,
+		KeyName:  group.Ec2SshKey,
+		UserData: group.UserData,
+		BlockDeviceMappings: []*ec2.LaunchTemplateBlockDeviceMappingRequest{
+			{
+				DeviceName: aws.String(storageDeviceName),
+				Ebs: &ec2.LaunchTemplateEbsBlockDeviceRequest{
+					VolumeSize: group.DiskSize,
+				},
+			},
+		},
+		TagSpecifications: utils.CreateTagSpecs(group.ResourceTags),
+	}
+	if !aws.BoolValue(group.RequestSpotInstances) {
+		launchTemplateData.InstanceType = group.InstanceType
+	}
+
+	return launchTemplateData
+}
+
+func newLaunchTemplateVersionIfNeeded(config *eksv1.EKSClusterConfig, upstreamNg, ng eksv1.NodeGroup, ec2Service *ec2.EC2) (*eksv1.LaunchTemplate, error) {
+	if aws.StringValue(upstreamNg.UserData) != aws.StringValue(ng.UserData) ||
+		aws.StringValue(upstreamNg.Ec2SshKey) != aws.StringValue(ng.Ec2SshKey) ||
+		aws.Int64Value(upstreamNg.DiskSize) != aws.Int64Value(ng.DiskSize) ||
+		aws.StringValue(upstreamNg.ImageID) != aws.StringValue(ng.ImageID) ||
+		(!aws.BoolValue(upstreamNg.RequestSpotInstances) && aws.StringValue(upstreamNg.InstanceType) != aws.StringValue(ng.InstanceType)) ||
+		!utils.CompareStringMaps(aws.StringValueMap(upstreamNg.ResourceTags), aws.StringValueMap(ng.ResourceTags)) {
+		lt, err := createNewLaunchTemplateVersion(config.Status.ManagedLaunchTemplateID, ng, ec2Service)
+		if err != nil {
+			return nil, err
+		}
+		config.Status.TemplateVersionsToDelete = append(config.Status.TemplateVersionsToDelete, aws.String(strconv.FormatInt(*upstreamNg.LaunchTemplate.Version, 10)))
+		return lt, nil
+	}
+
+	return nil, nil
+}
+
+func deleteLaunchTemplateVersions(templateID string, templateVersions []*string, ec2Service *ec2.EC2) {
+	launchTemplateDeleteVersionInput := &ec2.DeleteLaunchTemplateVersionsInput{
+		LaunchTemplateId: aws.String(templateID),
+		Versions:         templateVersions,
+	}
+
+	_, err := ec2Service.DeleteLaunchTemplateVersions(launchTemplateDeleteVersionInput)
+	if err != nil {
+		logrus.Warnf("could not delete version [%s] of launch template [%s]: %v, will not retry",
```

Comment:
I understand there are some issues with us returning errors here. Even if we only show the errors not related to a race, they would confusee the user and incorrectly imply that an update was not successful. Maybe we can at least retry here 3-5 times. I feel it is important we do our best to clean these up. I do worry that it will not be apparent if something is going on that prevents versions from being deleted so AWS ends up with a ton of them.

Reply:
Implemented a retry that will check for "no found" errors and retry deleting of template versions that don't have "not found" errors.